clippy_lints/src/must_use_self_return.rs

```diff
@@ -0,0 +1,78 @@
+use clippy_utils::{diagnostics::span_lint, must_use_attr, nth_arg, return_ty};
+use rustc_hir::intravisit::FnKind;
+use rustc_hir::{Body, FnDecl, HirId};
+use rustc_lint::{LateContext, LateLintPass};
+use rustc_session::{declare_lint_pass, declare_tool_lint};
+use rustc_span::Span;
+
+declare_clippy_lint! {
+    /// ### What it does
+    /// This lint warns when a method returning `Self` doesn't have the `#[must_use]` attribute.
+    ///
+    /// ### Why is this bad?
+    /// It prevents to "forget" to use the newly created value.
+    ///
+    /// ### Example
+    /// ```rust
+    /// pub struct Bar;
+    ///
+    /// impl Bar {
+    ///     // Bad
+    ///     pub fn bar(&self) -> Self {
+    ///         Self
+    ///     }
+    ///
+    ///     // Good
+    ///     #[must_use]
+    ///     pub fn foo(&self) -> Self {
+    ///         Self
+    ///     }
+    /// }
+    /// ```
+    #[clippy::version = "1.59.0"]
+    pub MUST_USE_SELF_RETURN,
+    style,
+    "Missing `#[must_use]` annotation on a method returning `Self`"
+}
+
+declare_lint_pass!(MustUseSelfReturn => [MUST_USE_SELF_RETURN]);
+
+impl<'tcx> LateLintPass<'tcx> for MustUseSelfReturn {
+    fn check_fn(
+        &mut self,
+        cx: &LateContext<'tcx>,
+        kind: FnKind<'tcx>,
+        decl: &'tcx FnDecl<'tcx>,
+        _: &'tcx Body<'tcx>,
+        span: Span,
+        hir_id: HirId,
+    ) {
+        if_chain! {
+            // We are only interested in methods, not in functions or associated functions.
```

Comment:
Why not? What about `fn new() -> Self`, `fn new_foo(bar: Baz) -> Self` or `fn from_blurb(blurb: Blurb) -> Self`?

Reply:
They look like different use cases to me. What I'm trying to prevent here is cases like `0f32.rad(2.)`. It's much less likely (in my opinion) if it's not a method.

Reply:
Ah sure, if you don't want to cover that case here that makes sense. I also think it's quite unlikely that someone calls such a constructor function and forgets to actually do something with the result. Adding `#[must_use]` on all these might add quite a bit of noise to the code.